precompile/contracts/ibc/contract.go

```diff
@@ -0,0 +1,145 @@
+package ibc
+
+import (
+	_ "embed"
+	"errors"
+	"fmt"
+	"math/big"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ibctransfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	ibcclienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	ethermint_statedb "github.com/evmos/ethermint/x/evm/statedb"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/precompile/contract"
+	"github.com/ethereum/go-ethereum/vmerrs"
+)
+
+const (
+	transferGasCost uint64 = contract.WriteGasCostPerSlot
+)
+
+// Singleton StatefulPrecompiledContract.
+var (
+	// IBCRawABI contains the raw ABI of IBC contract.
+	//go:embed IBC.abi
+	IBCRawABI string
+
+	IBCABI = contract.MustParseABI(IBCRawABI)
+
+	IBCPrecompile = createIBCPrecompile()
+)
+
+var ErrUnsupportedStateDB = errors.New("unsupported statedb")
+
+func transfer(
+	accessibleState contract.AccessibleState,
+	caller common.Address,
+	addr common.Address,
+	input []byte,
+	suppliedGas uint64,
+	readOnly bool,
+) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = contract.DeductGas(suppliedGas, transferGasCost); err != nil {
+		return nil, 0, err
+	}
+	if readOnly {
+		return nil, remainingGas, vmerrs.ErrWriteProtection
+	}
+
+	// Set the nonce of the precompile's address (as is done when a contract is created) to ensure
+	// that it is marked as non-empty and will not be cleaned up when the statedb is finalized.
+	accessibleState.GetStateDB().SetNonce(ContractAddress, 1)
+	// Set the code of the precompile's address to a non-zero length byte slice to ensure that the precompile
+	// can be called from within Solidity contracts. Solidity adds a check before invoking a contract to ensure
+	// that it does not attempt to invoke a non-existent contract.
+	accessibleState.GetStateDB().SetCode(ContractAddress, []byte{0x1})
```

Comment:
Should we move this to the geth [`contract.Run`](https://github.com/Kava-Labs/go-ethereum/blob/935f656757169f24b08b9e79a6371c1202a1fa8e/precompile/contract/contract.go#L70-L94) parent method?

Since it seems it would be shared logic for all stateful precompiles and can be covered by the parent test instead of for each individual precompile.